src/TLabel.cpp

```diff
@@ -104,7 +124,20 @@ void TLabel::mousePressEvent(QMouseEvent* event)
 
 void TLabel::mouseDoubleClickEvent(QMouseEvent* event)
 {
-    static_cast<void>(forwardEventToMapper(event));
+    if (forwardEventToMapper(event)) {
+        return;
+    }
+    if (mMouseButtons.contains(event->button())) {
+        if (mpHost) {
+            TEvent tmpDoubleClickParams = mDoubleClickParams;
+            tmpDoubleClickParams.mArgumentList.append(mMouseButtons.value(event->button()));
+            tmpDoubleClickParams.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+            mpHost->getLuaInterpreter()->callEventHandler(mDoubleClick, tmpDoubleClickParams);
```

Comment:
Do we also want a lua _table_ containing the _other_ buttons that are depressed at this time (in `event->buttons()`)?

Reply:
Sure! I had the same thought, especially in the mouseMove case, but honestly didn't know how to push a Lua table onto the stack. I'll look into it!

Reply:
Have a look where you see a `lua_newtable(L)` call - that _starts_ a new table, then you push a key (either an integer index - **starting at 1 (and not 0 for lua!)** for a an index table/list or a string/number/possibly some other more exotic lua data types for an associative table/array) and then a value (or start a sub-table with another `lua_newtable(L)` and then pop the pair of items off the _stack_ where you pushed them into the table with `lua_settable(L, -3)` - e.g. for a simple example from the bottom of `TLuaInterpreter::getDoors(...)`

    lua_newtable(L);
    QStringList keys = pR->doors.keys();
    for (unsigned int i = 0, total = keys.size(); i < total; ++i) {
        lua_pushstring(L, keys.at(i).toUtf8().constData());
        lua_pushnumber(L, pR->doors.value(keys.at(i)));
        lua_settable(L, -3);
    }
    return 1;

note that is using (unique) strings to make an associative table, the other form that uses a monotonically increment set of integers (i.e. goes 1, 2, 3, 4, ...n _without any gaps_) is effectively a list (which can be iterated through in lua with the `for index, value in ipairs(table) do ... end` construct and which can have the correct total of elements counted with the `#` operator prefixed on the `table` name - importantly there must NOT be any gaps in the sequence of index keys)

Finally as the table is a __single__ item that is being returned we `return 1` at the end of the function.

_\<aside>When I started working with Lua in C++ code I did it in something like a [Cargo cult programming](https://en.wikipedia.org/wiki/Cargo_cult_programming) mode - and there is plenty that I do not fully grasp even now - but there are other coders here who have a much more insightful understanding of the beast who will I am sure chip in if necessary. 😰 \</aside>_

Reply:
Let's not have humongous events that have all sorts of information in them that goes unused - this would e better served as a function that people can call if they want to see other buttons that are pressed at the time.

This shouldn't be an indexed table either but a key-value one.

Reply:
Hang on - the whole idea of __events__ is that they are something that is fired off that other things can listen for and respond to _if wanted_.  If some one wants to check that ONLY one mouse button was depressed at the time of the event (in this case the double click one) the details about the other buttons MUST be captured _at the same time_ - so it is eminently desirable to include that information even if some/most users will not care, so that those that do _can_ get it - besides I do not see a straightforward way to "store" the information for subsequent retrieval.

Should my view be considered reasonable the first value added to the parameters stored will be the mouse button that caused the (double click) event, and the second value added will be a list (indexed table) of the other button names that are also depressed at the same time (exactly reproducing the Qt library data).  This construct to be added to all the other mouse events (so that the move one indicates all the buttons down during the move) - this make a little more sense that a key-value table as the second value where the keys are ALL the mouse buttons the system knows about and a bool value of whether they are depressed.  Apart from anything else: that might be a _tiny_, _little bit_ confusing when talking about the mouse button release event unless it is clear that the released button must by definition be a `false` in the _other buttons down_ list; AND a lot more data (27 mouse buttons) to report the state of!

Can I also quietly mention _Qt::KeyboardModifiers_ and that they too ought to go into the mix? 😨

Reply:
Profiles are single-threaded, inside the event the buttons are still depressed and you can get that information if you want. It is not reasonable to have an event with 27 arguments attached to it!

Reply:
We're only talking about a single additional argument, I think... thought that argument is a Lua table with *potentially* 27 members. And if its the final argument being passed, you won't even have to acknowledge its existence unless you want the information it contains.

Personally, I'd be a fan of passing all information in the relevant QEvent on to the Lua interpreter... The more powerful the interface, the more cool stuff we'll see users make out of it. Just my opinion, though.

Reply:
I agree with WW - if either of you have ever used Linux `gpm` to force a reboot when the keyboard has stopped working (_triple left+right_ click followed by a _right_ click to do a `shutdown -r now`) having more then a couple of mouse buttons down at the same time is highly unlikely to be that common place (unless you have non-human digital appendages or some really exotic mouse hardware) which suggests just listing them in a table/list is the simplest data to pass, plus a third one (or possibly we could include them as well in the same table) for the keyboard modifiers (although merging the two together would depend on how the control/shift/alt and other modifiers are already encoded for some other lua functions added in the fairly recent past which we ought to reuse if we can).

Reply:
Right, I misunderstood the intentions. Sounds good to me then.

Reply:
So would it be okay if, instead of what I've done in this PR, I push a Lua key-value table as a single argument, containing all the information in QEvent, QMouseEvent, or QWheelEvent? Essentially, passing all information the QWidget::mousePressEvent, mouseReleaseEvent, mouseDoubleClickEvent, etc receives on to the Lua interpreter?

Reply:
Well, that wouldn't work, because then what's the point of having separate Mudlet events at all if everything comes as an argument? You'll be forcing the user to do all of the event matching inside the handler, making them replicate what they could have easily done just by hooking into individual event.

Let's add what other things are currently pressed but keep the events separate

Reply:
Sorry, I meant keep the events separate, just pass everything that a particular event receives within the Qt framework on to the matching Lua event. For example, all info in the QMouseEvent that QWidget::mousePressEvent  receives gets passed on to to the setLabelClickCallback event. All info in the QWheelEvent that QWidget::wheelEvent receives gets passed on to the setLabelWheelCallback event. And so on. 

In other words, when a label receives an event, don't pick out pieces of information, pass the whole thing on to the appropriate Lua event handler.

Reply:
Not sure we need everything, some stuff in there isn't relevant for scripting and would be confusing.